### PR TITLE
86c039c8g Implement JobApplicationNotes

### PR DESCRIPTION
**Pull request by gefasim**
Contains DB and API changes

1. Added new table - `job_application_notes`
2. Added new CRUD endpoints for JobApplicationNotes
3. JobApplicationNotes added into GET `/job-applications` endpoints

### CODE REVIEW

File: src/utils/array.extensions.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    return this.length !== uniqueSet.size;
+  };
+  Array.prototype.equals = function <T>(this: Array<T>, arrayTarget: Array<T>) {
+    const set1 = new Set(this);
+    const set2 = new Set(arrayTarget);
+    return set1.size === set2.size && [...set1].every((x) => set2.has(x));
+  };
+}
+
+export {};
```
Comment (akucintavalent) [inline src/utils/array.extensions.ts:20]: I think this is not necessary.

File: src/utils/array.extensions.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+if (!Array.prototype.containsDuplicates) {
+  Array.prototype.containsDuplicates = function <T>(this: Array<T>) {
+    const uniqueSet = new Set(this);
+    return this.length !== uniqueSet.size;
+  };
+  Array.prototype.equals = function <T>(this: Array<T>, arrayTarget: Array<T>) {
+    const set1 = new Set(this);
+    const set2 = new Set(arrayTarget);
+    return set1.size === set2.size && [...set1].every((x) => set2.has(x));
+  };
```
Comment (akucintavalent) [inline src/utils/array.extensions.ts:17]: I don't think you need to pass `this` as an argument to the function it must be accessible inside the function already.

File: src/modules/job-applications/job-applications.service.ts [hunk 3]
```diff
@@ -19,18 +19,21 @@ export class JobApplicationsService {
...
+    const boardColumnExists = await this.boardColumnsRepository.existsBy({
+      id: columnId,
+      board: { user: { id: userId } },
     });
 
-    const boardId = boardColumn.board.id;
-    const boardExists = await this.boardsRepository.existsBy({ id: boardId, user: { id: userId } });
-    if (!boardExists) {
+    if (!boardColumnExists) {
       throw new BadRequestException("Board column doesn't exists");
```
Comment (akucintavalent) [inline src/modules/job-applications/job-applications.service.ts:28]: exist

File: src/modules/job-applications/job-applications.mapper.ts [hunk 4]
```diff
@@ -13,6 +15,7 @@ export class JobApplicationMapper {
     dto.columnId = entity.column?.id;
     dto.createdAt = entity.createdAt;
     dto.updatedAt = entity.updatedAt;
```
Comment (akucintavalent) [inline src/modules/job-applications/job-applications.mapper.ts:17]: In such places you can use lodash library, and you can do something like

```ts
Object.assign(dto, _.pick(entity, ['id', 'title', 'companyName', 'description', 'createdAt', 'updatedAt']))
```

Comment (akucintavalent) [reply src/modules/job-applications/job-applications.mapper.ts:17]: Although to be honest i don't really understand why we are having these mappers :D

File: src/modules/job-application-notes/job-application-notes.service.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    });
+
+    const noteEntitiesIds = noteEntities.map((x) => x.id);
+    this.validateRearrange(noteIds, noteEntitiesIds);
+
+    for (let i = 0; i < noteEntitiesIds.length; i++) {
+      noteEntities.find((x) => x.id === noteIds[i]).order = i;
+    }
+
+    await this.jobApplicationNotesRepository.upsert(noteEntities, ['id']);
```
Comment (akucintavalent) [inline src/modules/job-application-notes/job-application-notes.service.ts:78]: I don't understand how this works

File: src/modules/job-application-notes/job-application-notes.service.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    await this.jobApplicationNotesRepository.softDelete({ id });
+  }
+
+  private async validateJobApplication(jobApplicationId: string, userId: string) {
+    if (!jobApplicationId) {
+      throw new ArgumentInvalidException('jobApplicationId is required');
+    }
+
+    const isJobApplicationExists = await this.jobApplicationsRepository.existsBy({
```
Comment (akucintavalent) [inline src/modules/job-application-notes/job-application-notes.service.ts:99]: please remove `is` from the name

File: src/modules/job-application-notes/job-application-notes.service.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      id: jobApplicationId,
+      column: { board: { user: { id: userId } } },
+    });
+
+    if (!isJobApplicationExists) {
+      throw new BadRequestException("JobApplication doesn't exists");
```
Comment (akucintavalent) [inline src/modules/job-application-notes/job-application-notes.service.ts:105]: `exist`, without `s`

File: src/modules/job-application-notes/job-application-notes.mapper.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    return entity;
+  }
+
+  toDto(entity: JobApplicationNote): JobApplicationNoteDto {
+    const dto = new JobApplicationNoteDto();
+    dto.id = entity.id;
+    dto.content = entity.content;
+    dto.order = entity.order;
+    dto.createdAt = entity.createdAt;
+    dto.updatedAt = entity.updatedAt;
```
Comment (akucintavalent) [inline src/modules/job-application-notes/job-application-notes.mapper.ts:27]:
```ts
Object.assign(dto, _.pick(entity, ['id', 'content', 'order', 'createdAt', 'updatedAt']))
```

File: src/modules/job-application-notes/job-application-notes.controller.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    return this.jobApplicationNotesService.find(jobApplicationId, userId);
+  }
+
+  @Post()
+  @ApiOperation({ summary: 'Creates a new Note for a JobApplication' })
+  @ApiResponse({ status: 200, description: 'Note created', type: JobApplicationNoteDto })
+  @ApiResponse({ status: 400, description: 'Validation error' })
+  @ApiResponse({ status: 400, description: 'JobApplication not found' })
+  create(@Body() jobApplicationNote: CreateJobApplicationNoteDto, @AuthUser() user: AuthUserDto) {
+    return this.jobApplicationNotesService.create(jobApplicationNote, user.userId);
```
Comment (akucintavalent) [inline src/modules/job-application-notes/job-application-notes.controller.ts:30]: In other places you do `{ userId }: AuthUserDto`. Try to be a little bit more consistent

File: src/modules/job-application-notes/dtos/update-job-application-note.dto.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsOptional, IsString } from 'class-validator';
+
+export class UpdateJobApplicationNote {
+  @ApiProperty({ nullable: true })
+  @IsString()
+  @IsOptional()
+  content: string | null;
```
Comment (akucintavalent) [inline src/modules/job-application-notes/dtos/update-job-application-note.dto.ts:8]: I think `@IsOptional()` means you should use `content?: string | null`

**Review by akucintavalent — CHANGES_REQUESTED** [review]

File: src/utils/array.extensions.ts (hunk 1 above)
Comment (gefasim) [reply src/utils/array.extensions.ts:20]: It is required

File: src/utils/array.extensions.ts (hunk 2 above)
Comment (gefasim) [reply src/utils/array.extensions.ts:17]: I'm leaving for readability

File: src/modules/job-application-notes/job-application-notes.service.ts (hunk 5 above)
Comment (gefasim) [reply src/modules/job-application-notes/job-application-notes.service.ts:78]: upsert - Inserts a new entity or array of entities unless they already exist in which case they are updated instead
https://orkhan.gitbook.io/typeorm/docs/repository-api

**Review by akucintavalent — APPROVED** [review]

**Pull request merged**